pkg/kvstore/allocator/allocator.go

```diff
@@ -274,17 +275,24 @@ func (k *kvstoreBackend) GetByID(ctx context.Context, id idpool.ID) (allocator.A
 	return k.keyType.PutKey(string(s)), nil
 }
 
-// UpdateKey refreshes the record that this node is using this key -> id
-// mapping. When reliablyMissing is set it will also recreate missing master or
-// slave keys.
-func (k *kvstoreBackend) UpdateKey(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, reliablyMissing bool) error {
+// UpdateMasterKey refreshes the master record for the key -> id mapping. When reliablyMissing is set to false,
+// it will initially check if the value is correct. When recreating, it will do CreateOnly to ensure it does not overwrite
+// a different value.
+func (k *kvstoreBackend) UpdateMasterKey(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, reliablyMissing bool) error {
 	var (
 		err        error
-		recreated  bool
 		keyPath    = path.Join(k.idPrefix, id.String())
 		keyEncoded = []byte(k.backend.Encode([]byte(key.GetKey())))
 		valueKey   = path.Join(k.valuePrefix, k.backend.Encode([]byte(key.GetKey())), k.suffix)
 	)
+	if !reliablyMissing {
+		// If we expect the key to be present, and it has the expected value, exit early as it's already correct
+		// in the kvstore. If there is an error, we skip to the CreateOnly part.
+		val, err := k.backend.Get(ctx, keyPath)
+		if err == nil && val != nil && bytes.Equal(val, keyEncoded) {
```

Comment:
We can split this into a separate PR as well, since its not strictly required. (also reminds me that it deserves a separate commit maybe).

Main reason we want this is that we expect it to be present and up to date, and that is the case in 99.9% of the cases. A `Get` call here results in a single read-only `Range` request for a single key, while opening a transaction will always write the transaction to the WAL log and require an fsync on the etcd servers.

Reply:
We tested this internally, and the etcd load went down significantly.